src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -257,39 +281,66 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
             }
 
             // Because SemanticToken data is generated relative to the previous token it must be in order.
-            // We have a guarentee of order within any given language server, but the interweaving of them can be quite complex.
+            // We have a guarantee of order within any given language server, but the interweaving of them can be quite complex.
             // Rather than attempting to reason about transition zones we can simply order our ranges since we know there can be no overlapping range.
             newList.Sort();
 
             return newList;
         }
 
         // Internal and virtual for testing only
-        internal virtual async Task<IReadOnlyList<SemanticRange>?> GetCSharpSemanticRangesAsync(
+        internal virtual async Task<VersionedSemanticRange> GetCSharpSemanticRangesAsync(
             RazorCodeDocument codeDocument,
             TextDocumentIdentifier textDocumentIdentifier,
             Range? range,
             long? documentVersion,
-            CancellationToken cancellationToken)
+            CancellationToken cancellationToken,
+            string? previousResultId = null)
         {
-            var csharpResponses = await GetMatchingCSharpResponseAsync(textDocumentIdentifier, documentVersion, cancellationToken);
+            IReadOnlyList<int>? previousCSharpTokens = null;
+            if (previousResultId != null)
+            {
+                // Attempting to retrieve cached tokens for the C# generated document.
+                _csharpGeneratedSemanticTokensCache.TryGetValue(previousResultId, out previousCSharpTokens);
+            }
+
+            var csharpResponse = previousResultId == null || previousCSharpTokens == null
+                ? await GetMatchingCSharpResponseAsync(
+                    textDocumentIdentifier, documentVersion, cancellationToken)
+                : await GetMatchingCSharpEditsResponseAsync(
+                    textDocumentIdentifier, documentVersion, previousResultId, previousCSharpTokens, cancellationToken);
 
-            if (csharpResponses is null)
+            if (csharpResponse is null)
             {
-                return null;
+                return VersionedSemanticRange.Default;
             }
 
-            SemanticRange? previousSemanticRange = null;
             var razorRanges = new List<SemanticRange>();
-            for (var i = 0; i < csharpResponses.Data.Length; i += 5)
+            if (csharpResponse.ResultId is null)
             {
-                var lineDelta = csharpResponses.Data[i];
-                var charDelta = csharpResponses.Data[i + 1];
-                var length = csharpResponses.Data[i + 2];
-                var tokenType = csharpResponses.Data[i + 3];
-                var tokenModifiers = csharpResponses.Data[i + 4];
+                // Indicates no C# code in Razor doc.
+                return new VersionedSemanticRange(razorRanges, null);
```

Comment:
How does this differ from the scenario where we returned `VersionedSemanticRange.Default`? Passing null to ResultId seems suspicious, but if we've got good reasons for the difference we should  describe them a bit more.

Reply:
Returning `VersionedSemanticRange.Default` indicates that there was some issue with retrieving the C# response (e.g. no C# response or C# is out of sync). Returning `VersionedSemanticRange(razorRanges, null)` indicates the Razor doc contains no C#. I don't know if the latter is actually hit in practice, but it was needed so the existing semantic tokens tests passed.
Good suggestion to add better comments though! Just did that.